source/extensions/common/aws/credentials_provider_impl.cc

```diff
@@ -62,13 +72,85 @@ Credentials EnvironmentCredentialsProvider::getCredentials() {
   return Credentials(access_key_id, secret_access_key, session_token);
 }
 
-void MetadataCredentialsProviderBase::refreshIfNeeded() {
+void CachedCredentialsProviderBase::refreshIfNeeded() {
   const Thread::LockGuard lock(lock_);
   if (needsRefresh()) {
     refresh();
   }
 }
 
+bool CredentialsFileCredentialsProvider::needsRefresh() {
+  return api_.timeSource().systemTime() - last_updated_ > REFRESH_INTERVAL;
+}
+
+std::string GetEnvironmentVariableOrDefault(const std::string& variable_name,
+                                            const std::string& default_value) {
+  const char* value = getenv(variable_name.c_str());
+  return (value != NULL) && (value[0] != '\0') ? value : default_value;
+}
+
+void CredentialsFileCredentialsProvider::refresh() {
+  ENVOY_LOG(debug, "Getting AWS credentials from the credentials file");
+
+  const auto credentials_file = GetEnvironmentVariableOrDefault(
+      AWS_SHARED_CREDENTIALS_FILE, DEFAULT_AWS_SHARED_CREDENTIALS_FILE);
+  const auto profile = GetEnvironmentVariableOrDefault(AWS_PROFILE, DEFAULT_AWS_PROFILE);
+
+  extractCredentials(credentials_file, profile);
+}
+
+void CredentialsFileCredentialsProvider::extractCredentials(const std::string& credentials_file,
+                                                            const std::string& profile) {
+  if (credentials_file.empty() || profile.empty()) {
+    return;
+  }
+
+  std::ifstream file(credentials_file);
+  if (!file) {
+    ENVOY_LOG(error, "Error opening credentials file {}", credentials_file);
+    return;
+  }
+
+  std::string access_key_id, secret_access_key, session_token;
+  std::string profile_start = fmt::format("[{}]", profile);
+
+  bool found_profile = false;
+  std::string line;
+  while (std::getline(file, line)) {
+    if (line == profile_start) {
+      found_profile = true;
+      continue;
+    }
+
+    if (found_profile) {
+      // Stop reading once we find the start of the next profile.
+      if (line.rfind("[", 0) == 0) {
+        break;
+      }
+
+      std::vector<std::string> parts = absl::StrSplit(line, absl::MaxSplits('=', 1));
+      if (parts.size() == 2) {
+        auto upper = StringUtil::toUpper(parts[0]);
+        if (upper == AWS_ACCESS_KEY_ID) {
+          access_key_id = parts[1];
+        } else if (upper == AWS_SECRET_ACCESS_KEY) {
+          secret_access_key = parts[1];
+        } else if (upper == AWS_SESSION_TOKEN) {
+          session_token = parts[1];
+        }
+      }
+    }
+  }
+
+  ENVOY_LOG(debug, "Found following AWS credentials for profile '{}' in {}: {}={}, {}={}, {}={}",
+            profile, credentials_file, AWS_ACCESS_KEY_ID, access_key_id, AWS_SECRET_ACCESS_KEY,
+            secret_access_key.empty() ? "" : "*****", AWS_SESSION_TOKEN,
+            session_token.empty() ? "" : "*****");
+
+  cached_credentials_ = Credentials(access_key_id, secret_access_key, session_token);
```

Comment:
should we raise error if AWS_ACCESS_KEY_ID & AWS_SECRET_ACCESS_KEY are not found?

In the docs we are saying `The fields 'aws_access_key_id', 'aws_secret_access_key',
   and 'aws_session_token' must be defined for the profile in the credentials file for this method to succeed`. So we need to fails if minimum requirements are not met. I believe `AWS_ACCESS_KEY_ID` & `AWS_SECRET_ACCESS_KEY` are mandatory based on this line https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/aws/credentials_provider_impl.cc#L241 

```
    if (credentials.accessKeyId() && credentials.secretAccessKey()) {
      return credentials;
    }
```

Reply:
Updated the docs to make the behavior more clear.